python/pyspark/sql/tests/pandas/test_pandas_transform_with_state.py

```diff
@@ -1226,6 +1226,55 @@ def check_results(batch_df, batch_id):
                 checkpoint_path=checkpoint_path,
             )
 
+    def test_value_state_schema_evolution(self):
+        checkpoint_path = tempfile.mktemp()
+        input_path = tempfile.mktemp()
+
+        # First run with old schema
+        self._prepare_input_data(input_path + "/batch1.txt", ["k1"], ["v1"])
+
+        df = self._build_test_df(input_path)
+        output_schema = StructType([
+            StructField("id", StringType()),
+            StructField("value", StringType())
+        ])
+
+        query1 = df.groupBy("id").transformWithStateInPandas(
+            OldStateProcessor(),
+            output_schema,
+            "Update"
+        ).writeStream \
+            .format("memory") \
+            .queryName("evolution_test") \
+            .option("checkpointLocation", checkpoint_path) \
+            .start()
+
+        query1.processAllAvailable()
+        query1.stop()
+
+        # Verify first state format
+        results1 = self.spark.sql("SELECT * FROM evolution_test")
+        assert results1.collect() == [Row(id="k1", value="v1")]
+
+        # Second run with evolved schema
+        self._prepare_input_data(input_path + "/batch2.txt", ["k1"], ["v2"])
+
+        query2 = df.groupBy("id").transformWithStateInPandas(
+            NewStateProcessor(),
+            output_schema,
+            "Update"
+        ).writeStream \
+            .format("memory") \
+            .queryName("evolution_test") \
+            .option("checkpointLocation", checkpoint_path) \
+            .start()
+
+        query2.processAllAvailable()
+        query2.stop()
+
+        # Verify evolved state still works
```

Comment:
Can we add 1 test each for:
- upcast
- downcast
- remove field
- reorder field
- rename field ?